### PR TITLE
Add a check to enforce alwaysRun = true on test after-methods

### PR DESCRIPTION
**Pull request by ksobolew**
## Description

Turns out that the `@AfterX` methods do not get executed by default if any of the previous methods failed or were skipped - and that includes the test methods as well. So if any of the tests is skipped, cleanup for the test may also be skipped. The `alwaysRun = true` annotation enforces that the cleanup method always gets executed.

## Additional context and related issues

The failure to run the cleanup method may cause the resource leak detector to trigger and mask the actual cause the test was skipped.

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by hashhar** [comment]
That's expected and what all our listeners do. See `Listeners#reportListenerFailure`. Wouldn't we want to know about that instead of just logging a warning which no one would read?

**Comment by ksobolew** [comment]
That's definitely the goal to make it very clear (by failing the tests?). I was following existing listeners, some of which also throw an exception. I was expecting that it would be more visible.

**Comment by ksobolew** [comment]
I would say it's even less than ideal, because an exception in the listener makes it skip all tests :)

**Comment by ksobolew** [comment]
OK, now it works very loudly ;)

```
java.lang.RuntimeException: The @AfterX methods should have the alwaysRun = true attribute to make sure that they'll run even if tests were skipped:
public void io.trino.execution.BaseDataDefinitionTaskTest.tearDown()
	at io.trino.testng.services.ReportAfterMethodNotAlwaysRun.checkHasAfterMethodsNotAlwaysRun(ReportAfterMethodNotAlwaysRun.java:77)
	at io.trino.testng.services.ReportAfterMethodNotAlwaysRun.onBeforeClass(ReportAfterMethodNotAlwaysRun.java:55)
        ...


2023-03-28T13:13:36.632-0600	INFO	pool-24-thread-1	stderr	JVM will be terminated

[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  11:21 min
[INFO] Finished at: 2023-03-28T19:13:36Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M5:test (default-test) on project trino-main: There are test failures.
Error:  
Error:  Please refer to /home/runner/work/trino/trino/core/trino-main/target/surefire-reports for the individual test results.
Error:  Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
Error:  The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
Error:  Command was ...
Error:  Error occurred in starting fork, check output in log
Error:  Process Exit Code: 1
Error:  Crashed tests:
Error:  io.trino.operator.aggregation.TestApproximateSetGenericDouble
Error:  org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
Error:  Command was ...
Error:  Error occurred in starting fork, check output in log
Error:  Process Exit Code: 1
Error:  Crashed tests:
Error:  io.trino.operator.aggregation.TestApproximateSetGenericDouble
```

Thanks @hashhar

File: testing/trino-testing-services/src/test/java/io/trino/testng/services/TestReportAfterMethodNotAlwaysRun.java [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.AfterSuite;
+import org.testng.annotations.AfterTest;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import static org.assertj.core.api.Assertions.assertThatNoException;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class TestReportAfterMethodNotAlwaysRun
```
Comment (kokosing) [inline testing/trino-testing-services/src/test/java/io/trino/testng/services/TestReportAfterMethodNotAlwaysRun.java:27]: Can you please add a test cases where base test class has this incorrect annotation and implementation has no override and has override has override with annotation correct/incorrect? And also a case where base class has correct annotation.

Comment (ksobolew) [reply testing/trino-testing-services/src/test/java/io/trino/testng/services/TestReportAfterMethodNotAlwaysRun.java:27]: Right. Subclasses. Will look into it, thanks

File: testing/trino-testing-services/src/test/java/io/trino/testng/services/TestReportAfterMethodNotAlwaysRun.java [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+        @AfterTest(alwaysRun = true)
+        public void afterTest()
+        {
+        }
+    }
+
+    private static class TestNotAlwaysRun
+    {
+        @AfterClass(alwaysRun = true)
+        public void afterClass()
```
Comment (kokosing) [inline testing/trino-testing-services/src/test/java/io/trino/testng/services/TestReportAfterMethodNotAlwaysRun.java:270]: nit: ` public void afterClass() {}` everywhere, the file will be shorter

**Review by kokosing — APPROVED** [review]

File: testing/trino-testing-services/src/test/java/io/trino/testng/services/TestReportAfterMethodNotAlwaysRun.java (hunk 2 above)
Comment (ksobolew) [reply testing/trino-testing-services/src/test/java/io/trino/testng/services/TestReportAfterMethodNotAlwaysRun.java]: Done (but only methods, I left empty classes alone)

**Review by hashhar — APPROVED** [review]
looks good but I'll defer to @findepi as the listener expert.

File: testing/trino-testing-services/src/main/java/io/trino/testng/services/ReportAfterMethodNotAlwaysRun.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+public class ReportAfterMethodNotAlwaysRun
+        implements IClassListener
+{
+    private static final Set<AnnotationPredicate<?>> AFTER_ANNOTATIONS = ImmutableSet.of(
+            new AnnotationPredicate<>(AfterTest.class, not(AfterTest::alwaysRun)),
+            new AnnotationPredicate<>(AfterMethod.class, not(AfterMethod::alwaysRun)),
+            new AnnotationPredicate<>(AfterClass.class, not(AfterClass::alwaysRun)),
+            new AnnotationPredicate<>(AfterSuite.class, not(AfterSuite::alwaysRun)),
+            new AnnotationPredicate<>(AfterGroups.class, not(AfterGroups::alwaysRun)));
```
Comment (findepi) [inline testing/trino-testing-services/src/main/java/io/trino/testng/services/ReportAfterMethodNotAlwaysRun.java:49]: the predicate doesn't look for "after annotations". it looks for "after annotations that lack certain attribute"

AFTER_ANNOTATIONS -> VIOLATIONS?

Comment (ksobolew) [reply testing/trino-testing-services/src/main/java/io/trino/testng/services/ReportAfterMethodNotAlwaysRun.java]: Good idea, changed

**Review by Laonel — APPROVED** [review]

**Comment by kokosing** [comment]
Thank you @ksobolew

**Pull request merged**